File: LeetCode/213.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        int len = nums.size();
        if (len == 1) {
            return nums[0];
        }
        int dp[len+1][2];

        int ans = 0;
        memset(dp, 0, sizeof(dp));
        for (int i = 2; i <= len; i++) {
            dp[i][0] = max(dp[i - 1][0], dp[i - 1][1]);
            dp[i][1] = dp[i - 1][0] + nums[i - 1];
        }
        ans = max(dp[len][0], dp[len][1]);

        memset(dp, 0, sizeof(dp));
        for (int i = 1; i <= len; i++) {
            dp[i][0] = max(dp[i - 1][0], dp[i - 1][1]);
            dp[i][1] = dp[i - 1][0] + nums[i - 1];
        }
        ans = max(ans, dp[len][0]);

        return ans;
    }
};

File: LeetCode_go/213.go
package main

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func rob(nums []int) int {
	n := len(nums)
	if n == 1 {
		return nums[0]
	}
	if n == 2 {
		return max(nums[0], nums[1])
	}
	find := func(nums []int) int {
		a, b, ans := 0, 0, 0
		for _, num := range nums {
			a, b = b+num, max(a, b)
			ans = max(ans, max(a, b))
		}
		return ans
	}

	return max(find(nums[1:]), find(nums[2:n-1])+nums[0])
}
